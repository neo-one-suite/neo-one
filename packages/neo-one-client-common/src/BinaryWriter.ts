import { BN } from 'bn.js';
import { common, ECPoint, ECPointHex, InvalidFormatError, UInt160, UInt160Hex, UInt256, UInt256Hex } from './common';
import { utils } from './utils';

interface OpOptions<T> {
  readonly fn: (value: T, buffer: Buffer, pos: number) => void;
  readonly length: number;
  readonly value: T;
}

// tslint:disable-next-line: no-any
class Op<OpValue = any> {
  public readonly fn: (value: OpValue, buffer: Buffer, pos: number) => void;
  public readonly length: number;
  public readonly value: OpValue;
  private mutableNext?: Op;

  public constructor(options: OpOptions<OpValue>) {
    this.fn = options.fn;
    this.length = options.length;
    this.value = options.value;
  }

  public get next() {
    return this.mutableNext;
  }

  // tslint:disable-next-line: no-any
  public setNext<NextValue = any>(nextOp: Op<NextValue>) {
    if (this.mutableNext !== undefined) {
      throw new Error('next operation already set for this op.');
    }
    this.mutableNext = nextOp;
  }
}

const writeByte = (value: number, buffer: Buffer, position: number) => {
  // tslint:disable-next-line: no-object-mutation no-bitwise
  buffer[position] = value & 255;
};

const writeFromBuffer = (source: Buffer, target: Buffer, offset = 0) => {
  source.forEach((bit, index) => {
    // tslint:disable-next-line: no-object-mutation no-bitwise
    target[index + offset] = bit & 255;
  });
};

const writeUInt160 = (value: UInt160 | UInt160Hex, buffer: Buffer, position: number) => {
  const result = common.uInt160ToBuffer(value);
  if (result.length !== common.UINT160_BUFFER_BYTES) {
    throw new Error('Invalid UInt160');
  }
  writeFromBuffer(result, buffer, position);
};

const writeUInt256 = (value: UInt256 | UInt256Hex, buffer: Buffer, position: number) => {
  const result = common.uInt256ToBuffer(value);
  if (result.length !== common.UINT256_BUFFER_BYTES) {
    throw new Error('Invalid UInt256');
  }
  writeFromBuffer(result, buffer, position);
};

const writeECPoint = (value: ECPoint | ECPointHex, buffer: Buffer, position: number) => {
  const result = common.ecPointToBuffer(value);
  if (result.length !== common.ECPOINT_BUFFER_BYTES) {
    throw new Error('Invalid ECPoint');
  }

  writeFromBuffer(result, buffer, position);
};

export class BinaryWriter {
  private readonly head: Op;
  private mutableLength: number;
  private mutableTail: Omit<Op, 'next'>;

  public constructor() {
    this.head = new Op({
      fn: () => {
        // do nothing
      },
      length: 0,
      value: undefined,
    });
    this.mutableTail = this.head;
    this.mutableLength = 0;
  }

  private get length() {
    return this.mutableLength;
  }

  public toBuffer() {
    return this.finish();
  }

  public writeBytes(value: Buffer): this {
    return this.push({
      fn: writeFromBuffer,
      length: value.length,
      value,
    });
  }

  public writeUInt8(value: number): this {
    return this.push({
      fn: (val: number, buffer: Buffer, pos: number) => buffer.writeUInt8(val, pos),
      length: 1,
      value,
    });
  }

  public writeInt16LE(value: number): this {
    return this.push({
      fn: (val: number, buffer: Buffer, pos: number) => buffer.writeInt16LE(val, pos),
      length: 2,
      value,
    });
  }

  public writeUInt16LE(value: number): this {
    return this.push({
      fn: (val: number, buffer: Buffer, pos: number) => buffer.writeUInt16LE(val, pos),
      length: 2,
      value,
    });
  }

  public writeUInt16BE(value: number): this {
    return this.push({
      fn: (val: number, buffer: Buffer, pos: number) => buffer.writeUInt16BE(val, pos),
      length: 2,
      value,
    });
  }

  public writeInt32LE(value: number): this {
    return this.push({
      fn: (val: number, buffer: Buffer, pos: number) => buffer.writeInt32LE(val, pos),
      length: 4,
      value,
    });
  }

  public writeUInt32LE(value: number): this {
    return this.push({
      fn: (val: number, buffer: Buffer, pos: number) => buffer.writeUInt32LE(val, pos),
      length: 4,
      value,
    });
  }

  public writeInt64LE(value: BN): this {
    return this.push({
      fn: (val: BN, buffer: Buffer, pos: number) => {
        const source = val.toTwos(8 * 8).toArrayLike(Buffer, 'le', 8);
        writeFromBuffer(source, buffer, pos);
      },
      length: 8,
      value,
    });
  }

  public writeUInt64LE(value: BN): this {
    return this.push({
      fn: (val: BN, buffer: Buffer, pos: number) => {
        const source = val.toArrayLike(Buffer, 'le', 8);
        writeFromBuffer(source, buffer, pos);
      },
      length: 8,
      value,
    });
  }

  public writeBoolean(value: boolean): this {
    return this.push({
      fn: writeByte,
      length: 1,
      value: value ? 1 : 0,
    });
  }

  public writeUInt160(hash: UInt160 | UInt160Hex): this {
    return this.push({
      fn: writeUInt160,
      length: common.UINT160_BUFFER_BYTES,
      value: hash,
    });
  }

  public writeUInt256(hash: UInt256 | UInt256Hex): this {
    return this.push({
      fn: writeUInt256,
      length: common.UINT256_BUFFER_BYTES,
      value: hash,
    });
  }

  public writeFixed8(value: BN): this {
    return this.writeInt64LE(value);
  }

  public writeFixedString(value: string, length: number): this {
    if (value.length > length) {
      throw new InvalidFormatError('String too long');
    }
    const stringBuffer = Buffer.from(value, 'utf8');
    if (stringBuffer.length > length) {
      throw new InvalidFormatError('String buffer length too long');
    }

    return this.push({
      fn: (val: Buffer, buffer: Buffer, pos: number) => {
        writeFromBuffer(val, buffer, pos);
        const padLength = val.length - length;
        if (padLength > 0) {
          writeFromBuffer(Buffer.alloc(padLength, 0), buffer, pos + val.length);
        }
      },
      length,
      value: stringBuffer,
    });
  }

  public writeArray<T>(values: readonly T[], write: (value: T) => void): this {
    this.writeVarUIntLE(values.length);
    values.forEach(write);

    return this;
  }

  public writeObject<T, K extends keyof T>(value: T, write: (key: K, value: T[K]) => void): this {
    const entries = Object.entries(value) as Array<[K, T[K]]>;
    this.writeVarUIntLE(entries.length);
    entries.forEach(([key, val]) => {
      write(key, val);
    });

    return this;
  }

  public writeVarBytesLE(value: Buffer): this {
    this.writeVarUIntLE(value.length);

    return this.push({
      fn: (val: Buffer, buffer: Buffer, pos: number) => {
        writeFromBuffer(val, buffer, pos);
      },
      length: value.length,
      value,
    });
  }

  public writeVarBytesLEWithoutVar(value: Buffer): this {
    return this.push({
      fn: (val: Buffer, buffer: Buffer, pos: number) => {
        writeFromBuffer(val, buffer, pos);
      },
      length: value.length,
      value,
    });
  }

  public writeVarUIntLE(valueIn: number | BN): this {
    const value = new BN(valueIn);
    if (value.lt(utils.ZERO)) {
      throw new InvalidFormatError('Expected value to be zero or positive');
    }

    if (value.lt(utils.FD)) {
      this.writeUInt8(value.toNumber());
    } else if (value.lte(utils.FFFF)) {
      this.writeUInt8(0xfd);
      this.writeUInt16LE(value.toNumber());
    } else if (value.lte(utils.FFFFFFFF)) {
      this.writeUInt8(0xfe);
      this.writeUInt32LE(value.toNumber());
    } else {
      this.writeUInt8(0xff);
      this.writeUInt64LE(value);
    }

    return this;
  }

  public writeVarString(value: string, max?: number): this {
    let buffer = Buffer.from(value, 'utf8');
    if (max !== undefined) {
      buffer = buffer.slice(0, max);
    }

    return this.writeVarBytesLE(buffer);
  }

  public writeVarStringWithoutVar(value: string, max?: number): this {
    let buffer = Buffer.from(value, 'utf8');
    if (max !== undefined) {
      buffer = buffer.slice(0, max);
    }

    return this.writeVarBytesLEWithoutVar(buffer);
  }

  public writeECPoint(value: ECPoint): this {
    if (common.ecPointIsInfinity(value)) {
      return this.push({
        fn: writeByte,
        length: 1,
        value: common.ECPOINT_INFINITY_BYTE,
      });
    }

    return this.push({
      fn: writeECPoint,
      length: common.ECPOINT_BUFFER_BYTES,
      value,
    });
  }

  private finish() {
    const computedBuffer = Buffer.alloc(this.length);
    let head = this.head.next;
    let position = 0;
    // tslint:disable-next-line: no-loop-statement
    while (head !== undefined) {
      head.fn(head.value, computedBuffer, position);
      position += head.length;
      head = head.next;
    }

    return computedBuffer;
  }

  private push<T>(op: Omit<OpOptions<T>, 'next'>): this {
    const newOp = new Op(op);
    this.mutableTail.setNext(newOp);
    this.mutableTail = newOp;
    this.mutableLength += op.length;

    return this;
  }
}
