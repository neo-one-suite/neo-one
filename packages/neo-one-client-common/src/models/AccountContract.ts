import _ from 'lodash';
import { common, ECPoint, UInt160 } from '../common';
import { crypto } from '../crypto';
import { scriptHashToAddress } from '../helpers';
import { JSONHelper } from '../JSONHelper';
import { utils } from '../utils';
import {
  ContractParameterTypeModel,
  toContractParameterType,
  toJSONContractParameterType,
} from './ContractParameterTypeModel';
import { AccountContractJSON } from './types';

export interface AccountContractAdd {
  readonly script: Buffer;
  readonly parameterList: readonly ContractParameterTypeModel[];
  readonly scriptHash?: UInt160;
  readonly address?: string;
}
export class AccountContract {
  public static createMultiSigContract(m: number, publicKeys: readonly ECPoint[]): AccountContract {
    return new AccountContract({
      script: crypto.createMultiSignatureRedeemScript(m, publicKeys),
      parameterList: _.range(m).map(() => ContractParameterTypeModel.Signature),
    });
  }

  public static createSignatureContract(publicKey: ECPoint): AccountContract {
    return new AccountContract({
      script: crypto.createSignatureRedeemScript(publicKey),
      parameterList: [ContractParameterTypeModel.Signature],
    });
  }

  public static fromJSON(json: AccountContractJSON): AccountContract {
    return new AccountContract({
      script: JSONHelper.readBase64Buffer(json.script),
      parameterList: json.parameterlist.map(toContractParameterType),
      scriptHash: JSONHelper.readUInt160(json.scripthash),
      address: json.address,
    });
  }

  public readonly script: Buffer;
  public readonly parameterList: readonly ContractParameterTypeModel[];
  public readonly scriptHashInternal: () => UInt160;
  public readonly addressInternal: () => string;

  public constructor({ script, parameterList, scriptHash: scriptHashIn, address: addressIn }: AccountContractAdd) {
    this.script = script;
    this.parameterList = parameterList;
    this.scriptHashInternal = scriptHashIn ? () => scriptHashIn : utils.lazy(() => crypto.hash160(this.script));
    this.addressInternal = addressIn
      ? () => addressIn
      : utils.lazy(() => scriptHashToAddress(common.uInt160ToString(this.scriptHashInternal())));
  }

  public get scriptHash() {
    return this.scriptHashInternal();
  }

  public get address() {
    return this.addressInternal();
  }

  public serializeJSON(): AccountContractJSON {
    return {
      script: JSONHelper.writeBase64Buffer(this.script),
      parameterlist: this.parameterList.map(toJSONContractParameterType),
      scripthash: JSONHelper.writeUInt160(this.scriptHash),
      address: this.address,
    };
  }
}
