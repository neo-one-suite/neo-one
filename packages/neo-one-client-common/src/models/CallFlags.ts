export enum CallFlags {
  None = 0,
  ReadStates = 0b00000001,
  WriteStates = 0b00000010,
  AllowCall = 0b00000100,
  AllowNotify = 0b00001000,
  // tslint:disable-next-line: no-bitwise
  States = ReadStates | WriteStates,
  // tslint:disable-next-line: no-bitwise
  ReadOnly = ReadStates | AllowCall,
  // tslint:disable-next-line: no-bitwise
  All = States | AllowCall | AllowNotify,
}
