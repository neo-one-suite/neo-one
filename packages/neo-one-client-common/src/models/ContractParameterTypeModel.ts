import { InvalidContractParameterTypeError, InvalidContractParameterTypeJSONError } from '../errors';
import { ContractParameterTypeJSON } from './types';

export enum ContractParameterTypeModel {
  Any = 0x00,
  Boolean = 0x10,
  Integer = 0x11,
  ByteArray = 0x12,
  String = 0x13,
  Hash160 = 0x14,
  Hash256 = 0x15,
  PublicKey = 0x16,
  Signature = 0x17,
  Array = 0x20,
  Map = 0x22,
  InteropInterface = 0x30,
  Void = 0xff,
}

export const isContractParameterType = (value: number): value is ContractParameterTypeModel =>
  // tslint:disable-next-line strict-type-predicates
  ContractParameterTypeModel[value] !== undefined;

export const assertContractParameterType = (value: number): ContractParameterTypeModel => {
  if (isContractParameterType(value)) {
    return value;
  }
  throw new InvalidContractParameterTypeError(value);
};

export const toJSONContractParameterType = (type: ContractParameterTypeModel): ContractParameterTypeJSON =>
  assertContractParameterTypeJSON(ContractParameterTypeModel[type]);

const isContractParameterTypeJSON = (value: string): value is ContractParameterTypeJSON =>
  // tslint:disable-next-line strict-type-predicates no-any
  ContractParameterTypeModel[value as any] !== undefined;

export const assertContractParameterTypeJSON = (value: string): ContractParameterTypeJSON => {
  if (isContractParameterTypeJSON(value)) {
    return value;
  }
  throw new InvalidContractParameterTypeJSONError(value);
};

export const toContractParameterType = (value: ContractParameterTypeJSON): ContractParameterTypeModel =>
  assertContractParameterType(ContractParameterTypeModel[value]);
