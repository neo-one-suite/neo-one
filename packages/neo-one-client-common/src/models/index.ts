export * from './AccountContract';
export * from './ContractParameterTypeModel';
export * from './SignerModel';
export * from './VerifyResultModel';
export * from './Serializable';
export * from './WitnessModel';
export * from './WitnessScopeModel';
export * from './transaction';
export * from './nep17';
export * from './types';
export * from './vm';
export * from './trigger';
export * from './NefFileModel';
export * from './CallFlags';
export * from './MethodTokenModel';
