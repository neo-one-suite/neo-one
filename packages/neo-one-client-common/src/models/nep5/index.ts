export * from './Nep5BalanceModel';
export * from './Nep5TransferModel';
export * from './Nep5TransferKeyModel';
export * from './Nep5BalanceKeyModel';
