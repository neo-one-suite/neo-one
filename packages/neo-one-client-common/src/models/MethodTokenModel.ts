import { BinaryWriter } from '../BinaryWriter';
import { common, UInt160 } from '../common';
import { IOHelper } from '../IOHelper';
import { utils } from '../utils';
import { CallFlags } from './CallFlags';
import { createSerializeWire, SerializableJSON, SerializableWire, SerializeWire } from './Serializable';
import { MethodTokenJSON } from './types';

export interface MethodTokenAdd {
  readonly hash: UInt160;
  readonly method: string;
  readonly paramCount: number;
  readonly hasReturnValue: boolean;
  readonly callFlags: CallFlags;
}

export class MethodTokenModel implements SerializableJSON<MethodTokenJSON>, SerializableWire {
  public get size() {
    return this.sizeInternal();
  }

  public readonly hash: UInt160;
  public readonly method: string;
  public readonly paramCount: number;
  public readonly hasReturnValue: boolean;
  public readonly callFlags: CallFlags;

  public readonly serializeWire: SerializeWire = createSerializeWire(this.serializeWireBase.bind(this));

  private readonly sizeInternal = utils.lazy(
    () =>
      IOHelper.sizeOfUInt160 +
      IOHelper.sizeOfVarString(this.method) +
      IOHelper.sizeOfUInt16LE +
      IOHelper.sizeOfBoolean +
      IOHelper.sizeOfUInt8,
  );

  public constructor({ hash, method, paramCount, hasReturnValue, callFlags }: MethodTokenAdd) {
    this.hash = hash;
    this.method = method;
    this.paramCount = paramCount;
    this.hasReturnValue = hasReturnValue;
    this.callFlags = callFlags;
  }

  public serializeWireBase(writer: BinaryWriter): void {
    writer.writeUInt160(this.hash);
    writer.writeVarString(this.method);
    writer.writeUInt16LE(this.paramCount);
    writer.writeBoolean(this.hasReturnValue);
    writer.writeUInt8(this.callFlags);
  }

  public serializeJSON(): MethodTokenJSON {
    return {
      hash: common.uInt160ToString(this.hash),
      method: this.method,
      paramcount: this.paramCount,
      hasreturnvalue: this.hasReturnValue,
      callflags: this.callFlags,
    };
  }
}
