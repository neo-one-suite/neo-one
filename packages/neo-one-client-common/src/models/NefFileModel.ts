import { BinaryWriter } from '../BinaryWriter';
import { crypto } from '../crypto';
import { IOHelper } from '../IOHelper';
import { JSONHelper } from '../JSONHelper';
import { utils } from '../utils';
import { MethodTokenModel } from './MethodTokenModel';
import { createSerializeWire, SerializableJSON, SerializableWire, SerializeWire } from './Serializable';
import { NefFileJSON } from './types';

export interface NefFileAdd {
  readonly compiler: string;
  readonly tokens: readonly MethodTokenModel[];
  readonly script: Buffer;
  readonly checkSum?: number;
}

export class NefFileModel implements SerializableJSON<NefFileJSON>, SerializableWire {
  public get size() {
    return this.sizeInternal();
  }

  public static readonly magic = 0x3346454e;
  public readonly compiler: string;
  public readonly tokens: readonly MethodTokenModel[];
  public readonly script: Buffer;
  public readonly checkSum: number;
  public readonly serializeWire: SerializeWire = createSerializeWire(this.serializeWireBase.bind(this));
  public readonly serializeForChecksum: SerializeWire = createSerializeWire(
    this.serializeBeforeChecksumBase.bind(this),
  );

  private readonly headerSize = utils.lazy(() => IOHelper.sizeOfUInt32LE + IOHelper.sizeOfFixedString(64));
  private readonly sizeInternal = utils.lazy(
    () =>
      this.headerSize() +
      IOHelper.sizeOfUInt16LE +
      IOHelper.sizeOfArray(this.tokens, (token) => token.size) +
      IOHelper.sizeOfUInt16LE +
      IOHelper.sizeOfVarBytesLE(this.script) +
      IOHelper.sizeOfUInt32LE,
  );

  public constructor({ compiler, tokens, script, checkSum }: NefFileAdd) {
    this.compiler = compiler;
    this.tokens = tokens;
    this.script = script;
    this.checkSum = checkSum ?? crypto.sha256(this.serializeForChecksum()).readUInt32LE(0);
  }

  public serializeHeader(writer: BinaryWriter): void {
    writer.writeUInt32LE(NefFileModel.magic);
    writer.writeFixedString(this.compiler, 64);
  }

  public serializeBeforeChecksumBase(writer: BinaryWriter): void {
    this.serializeHeader(writer);
    writer.writeUInt16LE(0);
    writer.writeArray(this.tokens, (token) => token.serializeWire());
    writer.writeUInt16LE(0);
    writer.writeVarBytesLE(this.script ?? Buffer.from([])); // TODO: script can be null?
  }

  public serializeWireBase(writer: BinaryWriter): void {
    this.serializeBeforeChecksumBase(writer);
    writer.writeUInt32LE(this.checkSum);
  }

  public serializeJSON(): NefFileJSON {
    return {
      magic: NefFileModel.magic,
      compiler: this.compiler,
      tokens: this.tokens.map((token) => token.serializeJSON()),
      script: JSONHelper.writeBase64Buffer(this.script),
      checksum: this.checkSum,
    };
  }
}
