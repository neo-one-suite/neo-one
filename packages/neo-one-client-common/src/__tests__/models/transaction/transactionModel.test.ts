import { BN } from 'bn.js';
import { keys } from '../../../__data__/keys';
import { BinaryWriter } from '../../../BinaryWriter';
import { SignerModel, TransactionModel, TransactionModelAdd } from '../../../models';

const testSender = keys[0].scriptHash;
const testPrivKey = keys[1].privateKey;

const optionsBuilder = ({
  version = 0,
  nonce = 0,
  systemFee = new BN(0),
  networkFee = new BN(0),
  validUntilBlock = 100000,
  sender = testSender,
  attributes = [],
  witnesses = [],
  signers = [],
  hash,
  script = Buffer.from([0x10]),
}: Partial<TransactionModelAdd>): TransactionModelAdd => ({
  version,
  nonce,
  systemFee,
  networkFee,
  validUntilBlock,
  sender,
  attributes,
  witnesses,
  signers,
  hash,
  script,
});

// tslint:disable-next-line:no-let
let testWriter: BinaryWriter = new BinaryWriter();

const resetWriter = () => {
  testWriter = new BinaryWriter();
};

describe('Transaction Model', () => {
  beforeEach(resetWriter);

  const systemFee = new BN(1);
  const networkFee = new BN(1);
  const script = Buffer.from([0x01]);

  test('Serialize Base - Unsigned', () => {
    const transaction = new TransactionModel({ ...optionsBuilder({ systemFee, networkFee, script }) });
    transaction.serializeUnsignedBase(testWriter);
    expect(testWriter.toBuffer()).toMatchSnapshot();
  });

  test('Serialize Base - With Cosigners', () => {
    const signer = new SignerModel({
      account: testSender,
      scopes: 'Global',
    });

    const cosignerTransaction = new TransactionModel({
      ...optionsBuilder({
        systemFee,
        networkFee,
        script,
        signers: [signer],
      }),
    });

    const serializeUnsigned = cosignerTransaction.serializeUnsigned();
    expect(serializeUnsigned).toBeDefined();
  });

  test('Cloned Transaction', () => {
    const transaction = new TransactionModel({ ...optionsBuilder({ systemFee, networkFee, script }) });
    const clonedTransaction = transaction.clone();

    expect(JSON.stringify(transaction)).toEqual(JSON.stringify(clonedTransaction));
  });

  test('Serialize Base - Signed', () => {
    const transaction = new TransactionModel({ ...optionsBuilder({ systemFee, networkFee, script }) });
    const signedTransaction = transaction.sign(testPrivKey);
    const witness = signedTransaction.witnesses[0];
    expect(witness).toBeDefined();
    const serializedUnsigned = signedTransaction.serializeUnsigned();
    const serializedSigned = signedTransaction.serializeWire();
    expect(serializedSigned).toEqual(Buffer.concat([serializedUnsigned, Buffer.from([0x01]), witness.serializeWire()]));
  });
});
