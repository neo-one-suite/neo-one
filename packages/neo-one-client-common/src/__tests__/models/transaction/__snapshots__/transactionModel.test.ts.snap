// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Transaction Model Serialize Base - Unsigned 1`] = `"000000000001000000000000000100000000000000a086010000000101"`;
